README.md
# Pizzeria #
## Version beta ##

# Fonctionnement #

Le but du jeu est de gérer votre propre pizzeria

# Quelques détail techniques #

Vous devez pour le moment installer le module "emoji" de python avant de lancer le jeu 
	(il vous suffit soit de l'installer selon la technique de votre choix, soit en ouvrant un terminal et en tapant "pip install emoji")
Si vous avez des carrés qui s'affichent à la place des émojis, n'ayez craintes, je m'occupe de faire une version sans émoji pour vous
Pour tout autre problème, vous pouvez me contacter par mail (greggydu75@gmail.com) ou via mon site, je vous répondrais volontier

# Quelques bugs encore existants #

Si vous allez trop loin de la pizza, je ne garanti pas l'état du jeu :P
C'est très facile d'arnaquer le jeu pour l'instant, mais ça, il ne faut pas le dire...

# A propos #


core/history.md
## Etapes en cours ##


## Etapes futures ##


## Etapes finies ##

[Fini] Faire un système pour sauvegarder
[Fini] Faire un menu de fin
[Fini] Système d'amélioration
[Fini] Voir c'est quoi le bordel avec le payement de la pizza
[Fini] Faire des couleurs mieux pour voir que la pizza est cuite
[Fini] Mettre a jour la notation pour les tomates
[Fini] Faire en sorte que la sauce tomate soit debitee
[Fini] Ajouter une touche pour faire cuire la pizza et une touche pour servir la pizza
[Fini] Ajouter un ecran permettant d'afficher quelques parametres (numero de jour, budget restant...)
[Fini] Faire des pourboires
[Fini] Faire une impossibilité d'être à découvert --> N'ajoute plus les ingredients
[Fini] Faire un écran permettant d'afficher les touches permettant d'ajouter des ingrédients
[Fini] Faire un deuxième écran pour afficher la commande
[Fini] Faire un système permettant d'évaluer la qualité d'une pizza
[Fini] Faire en sorte que le prix des ingrédients soit soustrait au budget du joueur
[Fini] Placement impossible d'un ingrédient sur un autre ingrédient
[Fini] Placement impossible des ingredients en dehors de la pizza
[Fini] Dessin de la pizza
[Fini] Base des diverses classes



## Améliorations ##

[Amelioration] Faire un peu plus d'ingrédients
[Amelioration] Possibilité d'améliorations du temps de cuisson des pizzas
[Amelioration] Possibilité de débloquer les ingrédients les uns après des autres
[Amelioration] Mise en place de niveau permettant de débloquer de nouvelles améliorations


core/Day.py
# Day managment file

# Global import
import math;
import random;
import time;

# Local import
import Client;

class Day:
	def __init__(self, restaurant, main_scr, command_scr, keys_scr, const_scr):
		self.restaurant = restaurant;
		self.number_of_client = abs(random.randint(round(math.log10(restaurant.xp)**2) - 2, round(math.log10(restaurant.xp)**2) + 2)) + 1;
		self.main_scr = main_scr;
		self.keys_scr = keys_scr;
		self.command_scr = command_scr;
		self.const_scr = const_scr;
		self.income_of_the_day = 0;
		self.outgoings_expense_of_the_day = 0;

	def start_day(self):
		for i in range(self.number_of_client):
			current_client = Client.Client(self.restaurant, self.main_scr, self.command_scr, self.keys_scr, self.const_scr);
			self.income_of_the_day += current_client.payement;
			self.outgoings_expense_of_the_day += current_client.outgoings_expense;
			self.main_scr.clear();
			self.command_scr.clear();
			self.keys_scr.clear();
			self.main_scr.refresh();
			self.keys_scr.refresh();
			self.command_scr.refresh();
			time.sleep(random.randint(0, 2));
		
		self.end_of_day()
	
	def end_of_day(self):
		self.restaurant.debit(10, self.const_scr); # Payement of the rent
		self.main_scr.clear();
		self.keys_scr.clear();
		self.command_scr.clear();
		
		self.main_scr.addstr(2, 2, "Dépense du jour");
		self.main_scr.addstr(3, 2, "  Ingrédients : {0}€".format(self.outgoings_expense_of_the_day));
		self.main_scr.addstr(4, 2, "  Loyer : {0}€".format(self.restaurant.rent));
		
		self.main_scr.addstr(5, 2, "Entrées du jour");
		self.main_scr.addstr(6, 2, "  Payement des pizzas : {0}€".format(self.income_of_the_day));

		self.main_scr.addstr(7, 2, "Nombre de clients aujourd'hui : {0}".format(self.number_of_client));

		self.main_scr.refresh();
		self.keys_scr.refresh();
		self.command_scr.refresh();

		self.main_scr.getkey();


core/main.py
# Global import
import curses;

# Local import
import master;
import Day;

def end_day():
	main_scr.resize(20, 80);
	choice = master.end_day_menu(main_scr);

	if (choice == '1'):
		restaurant.upgrade_menu(main_scr, constant_scr);
		end_day();
	
	elif (choice == '2'):
		main_scr.resize(20, 40);
		new_day();
	
	elif (choice == '3'):
		master.save_game(restaurant);
		end_day();
	
	elif (choice == '4'):
		return (0);
	


def new_day():
	day = Day.Day(restaurant, main_scr, command_scr, keys_scr, constant_scr);
	day.start_day();

	end_day();


def main(screen):
	"""Main function"""
	global restaurant, main_scr, command_scr, keys_scr, constant_scr;
	
	main_scr, command_scr, keys_scr, constant_scr = master.init(screen);

	if (master.start_menu(main_scr) == '1'):
		restaurant = master.load_game(main_scr);
	
	else:
		restaurant = master.new_game(main_scr);
	
	new_day();


curses.wrapper(main);



core/Client.py
# Client file
"""
Client documentation

Properties:
>>> int nb_pizza : "number of wanted pizza"
>>> window scr : "main screen"
>>> window command_scr : "screen that display the command of the client for the current pizza"
>>> window keys_scr : "screen that display connections between keys and ingredients"
>>> window constant_scr : "screen that show money left, level, number of day, ..."
>>> Restaurant restaurant : "restaurant object (the player)"
>>> list ingredient_choice : "ingredients that are wanted on the pizza"
>>> int price : "price of the pizza (without margin)"
>>> Pizza current_pizza : "pizza that is cooked"
>>> int payement : "price decided by the client"
>>> int outgoings_expense : "outgoings expense for every pizza of the client"

Methods:
>>> __init__(self, Restaurant restaurant, window scr window command_scr, window keys_scr, window constant_scr) : "same as properties"
	construtor of the client
>>> generate_command(self)
	generation of th command
>>> display_command(self)
	display the command and the connection between keys and ingredients
>>> served(self)
	success rate the pizza and pay the player acording to the success rate
"""
# Global import
import random;
import time;
import json;
import math;

# Local import
import Pizza;

class Client:
	def __init__(self, restaurant, scr, command_scr, keys_scr, constant_scr):
		self.nb_pizzas = 1; # Possible improvment
		self.scr = scr;
		self.command_scr = command_scr;
		self.keys_scr = keys_scr;
		self.constant_scr = constant_scr;
		self.restaurant = restaurant;
		self.generate_command();
		self.outgoings_expense = 0;
		self.display_command();
		self.restaurant.display_budget(self.constant_scr);
		self.restaurant.display_const(self.constant_scr);
		self.current_pizza = Pizza.Pizza(self, self.restaurant, self.ingredient_choice, self.scr, constant_scr); # Creation of the pizza object
		self.time_start = time.time();
		self.current_pizza.do_pizza();
	
	#===========================================================================================
	
	def generate_command(self):
		with open("../languages/french/orders.json", "r") as f:
			orders = json.load(f);

		order_choice = random.sample(orders.keys(), 1);
		
		ingredient_unavaible = False;
		for i in order_choice:
			if (i not in self.restaurant.ingredients):
				ingredient_unavaible = True;

		if (ingredient_unavaible):
			self.generate_command();
		else:
			self.ingredient_choice = "".join(order_choice).split(", ");

	#===========================================================================================

	def display_command(self):
		self.command_scr.clear();
		self.keys_scr.clear();
		with open("../languages/french/orders.json", "r") as f:
			command = json.load(f)[", ".join(self.ingredient_choice)];
			self.command_scr.addstr(0, 0, command);
		
		temp = list(self.restaurant.connection_key_ingredient_display.items());
		for i in range(len(self.restaurant.connection_key_ingredient) + 1):
			self.keys_scr.addstr(i, 0, temp[i][0] + " : " + temp[i][1]);

		self.command_scr.refresh();
		self.keys_scr.refresh();

	#===========================================================================================

	def served(self):
		waiting_time = self.time_start - time.time();
		ingredients_wanted_on_pizza = self.current_pizza.ingredients_on_pizza;
		unwanted_ingredients = 0;
		
		# Calculation of unwanted ingredients
		for elt, number in self.current_pizza.ingredients_on_pizza.items():
			if(not elt in self.ingredient_choice):
				unwanted_ingredients += number;
				del ingredients_wanted_on_pizza[elt];

		# If there are more than 8 ingredients of the same type, the other ingredients of that type are not taking into consideration
		temp = ingredients_wanted_on_pizza;
		for elt, number in temp.items():
			if (number > 8):
				ingredients_wanted_on_pizza[elt] = 8;

		# Calculation of the ingredients rate
		ingredients_success_rate = sum(ingredients_wanted_on_pizza.values()) / (len(self.ingredient_choice) * 8) * 100;
		ingredients_success_rate -= unwanted_ingredients * 2;

		# Calculation of cooking rate
		cooking_rate = 1 if self.current_pizza.cooked else 0.90

		# Calculation of waiting rate
		if (waiting_time >= 10):
			waiting_rate = 1.10;

		elif (waiting_time >= 25):
			waiting_rate = 1;
		
		elif (waiting_time >= 40):
			waiting_rate = 0.90;
		
		elif (waiting_time >= 55):
			waiting_rate = 0.80;
		
		elif (waiting_time >= 60):
			waiting_rate = 0.70;
		
		else:
			waiting_rate = 0.50;

		# Calculation of the global rate
		global_success_rate = ingredients_success_rate * cooking_rate * waiting_rate;

		with open("../packages/all_ingredients.json", "r") as f:
			prices = json.load(f);
			self.minimum_price = sum([prices[i] * ingredients_wanted_on_pizza[i] for i in ingredients_wanted_on_pizza.keys()]) + 2;
		
		self.payement = round(self.minimum_price * (1 * (global_success_rate/100)), 2);
		self.restaurant.credit(self.payement, self.constant_scr);
		self.outgoings_expense += self.current_pizza.outgoings_expense;
		self.restaurant.xp += self.payement;
		self.restaurant.display_const(self.constant_scr);



core/master.py
# Main file of the pizzeria
"""
Colors numbers:
0: BLACK
1: RED
2: GREEN
3: YELLOW
4: BLUE
5: MAGENTA
6: CYAN
7: WHITE
8: LIGHT YELLOW
9: DARK YELLOW
10: ???

Colors pairs:
0: WHITE / BLACK
1: RED / BLACK
2: YELLOW / BLACK
3: LIGHT YELLOW / BLACK
4: DARK YELLOW / BLACK
"""
# Global import
import curses;
import os;
import pickle;

# Local import
import Restaurant;

def init(main_screen):
	"""Main function"""
	# Colors init #
	curses.init_color(8, 700, 500, 0);
	curses.init_color(9, 999, 800, 0);
	curses.init_pair(1, curses.COLOR_RED, curses.COLOR_BLACK);
	curses.init_pair(3, 8, curses.COLOR_BLACK);
	curses.init_pair(4, 9, curses.COLOR_BLACK);
	curses.curs_set(0);
	main_scr = curses.newwin(20, 40, 9, 0); # The main self.main_screen
	command_scr = curses.newwin(3, 60, 5, 0); # The command self.main_screen
	keys_scr = curses.newwin(20, 40, 9, 40); # self.main_screen that display connection between key and ingredient
	constant_scr = curses.newwin(4, 60, 0, 0); # self.main_screen that display money, day, etc...
	main_scr.keypad(True);

	return(main_scr, command_scr, keys_scr, constant_scr);

def start_menu(main_scr):
	"""Function that manage the start menu"""
	main_scr.clear();

	main_scr.addstr(3, 5, "Menu principal");
	main_scr.addstr(4, 5, "1 : Charger une partie");
	main_scr.addstr(5, 5, "2 : Nouvelle partie");
	
	key = '100';
	while(key not in '12'):
		key = main_scr.getkey();
	
	return(key);

#================================================================

def end_day_menu(main_scr):
	main_scr.clear();
	
	main_scr.addstr(3, 2, "Que voulez-vous faire ?");
	main_scr.addstr(4, 2, "1 : Améliorer votre restaurant (pas dispo)");
	main_scr.addstr(5, 2, "2 : Ouvrir la pizzeria le lendemain");
	main_scr.addstr(6, 2, "3 : Sauvegarder votre partie (rien ne se passe, c'est normal)");
	main_scr.addstr(7, 2, "4 : Quitter (sans sauvegarder)");

	key = '100';
	while(key not in '1234'):
		key = main_scr.getkey();
	
	main_scr.clear();
	main_scr.refresh();

	return(key);

#=================================================================

def existing_game(game):
	"""str game --> bool
	Check is the game \'game\' already exist or not"""
	
	return(os.path.exists("../games/" + game));

#================================================================

def load_game(main_scr):
	"""Funciton that will load an existing game"""
	main_scr.clear();
	curses.curs_set(2); # Show the cursor
	curses.echo();
	main_scr.addstr(3, 5, "Quel est votre nom ? ");
	name = main_scr.getstr().decode("utf-8");

	while(not existing_game(name)):
		main_scr.clear();
		main_scr.addstr(3, 5, "Cette sauvegarde n'existe pas... (Veillez à respecter la casse)");
		main_scr.addstr(4, 5, "Entrez le nom d'une sauvegarde : ");
		name = main_scr.getstr().decode("utf-8");
		
	curses.noecho();
	curses.curs_set(0); # Hide the cursor

	with open("../games/" + name + "/restaurant.save", "rb") as restaurant_file: # Get the Player object
		restaurant_ = pickle.load(restaurant_file);
	
	return(restaurant_);

#================================================================

def save_game(restaurant):
	os.mkdir("../games/" + restaurant.name);
	with open("../games/" + restaurant.name + "/restaurant.save", "wb") as f:
		pickle.dump(restaurant, f);

#=================================================================

def new_game(main_scr):
	"""Function that create a new game"""
	curses.curs_set(2); # Show the cursor
	main_scr.clear(); # Clear the main_screen
	curses.echo(); # Display what the user is writting

	main_scr.addstr(3, 5, "Quel est votre prénom ? ");
	surname = main_scr.getstr().decode("utf-8");
	main_scr.addstr(4, 5, "Quel est votre nom ? ");
	name = main_scr.getstr().decode("utf-8");

	while(existing_game(name)): # Check if the game already exists
		main_scr.clear();
		main_scr.addstr(3, 5, "Cette partie existe déjà...")
		main_scr.addstr(4, 5, "Quel est votre prénom ? ");
		surname = main_scr.getstr().decode("utf-8");
		main_scr.addstr(5, 5, "Quel est votre nom ? ");
		name = main_scr.getstr().decode("utf-8");
		
	curses.noecho(); # Hide typed character
	curses.curs_set(0); # Hide the cursor

	restaurant_ = Restaurant.Restaurant(name, surname);

	return(restaurant_);



core/Restaurant.py
# Restaurant manage file
"""
Restaurant documentation

Properties: 
>>> list ingredients --> "unlocked ingredients"
>>> int cooking_time --> "time of waiting for cooking a pizza"
>>> dict connecion_key_ingredient --> "connections between a key and an ingredient (only for the computer)"
>>> dict connection_key_ingredient_display --> "same as before but for displaying on the screen" 
>>> str name --> "name of the player"
>>> str surname --> "surname of the player"
>>> int xp --> "total xp number"
>>> int level --> "level of the restaurant"
>>> float money --> "money left for the player"
>>> int days --> "number of days passed since the start of the game"
>>> int rent --> "rent for a day"

Methods:
>>> __init__(self, str name, str surname) --> "name --> name of the player, surname --> surname of the player"
	 constructor of the restaurant
>>> credit(self, float how_much, curses_screen constant_scr) --> "how_much --> how much credit, constant_scr --> screen of money, days, level, ..."
	 credit "how much" to the money of the player
>>> debit(self, float how_much, curses_screen constant_scr) --> "same as previous line"
	 debit "how_much" of to money of the player
>>> display_budget(self, curses_screen constant_scr) --> "constant_scr --> screen of money, days, level, ..."
	 display player budget		
"""
# Global import
import json;
import curses;
import emoji;

class Restaurant:
	def __init__(self, name, surname):
		with open("../packages/default_ingredients.json", "r") as f:
			self.ingredients = json.load(f);

		with open("../packages/default_cooking_time.txt", "r") as f:
			self.cooking_time = int(f.read().replace("\n", ""));
			self.oven_level = 1;

		with open("../packages/default_connection_key_ingredient.json", "r") as f:
			self.connection_key_ingredient = json.load(f); # {'key': 'emoji_code_of_ingredient', ...}

		with open("../languages/french/default_connection_key_ingredient_display.json", "r") as f:
			self.connection_key_ingredient_display = json.load(f); # {'key': 'name_of_ingredient', ...}
			
		self.name = name;
		self.surname = surname;
		self._xp = 1;
		self.level = 1;
		self._money = 50;
		self.days = 1;
		self.rent = 10;


	@property
	def xp(self):
		return(self._xp);

	@xp.setter
	def xp(self, new_xp):
		self._xp = new_xp;
		self._xp = round(self._xp, 0);
		
		if self._xp - 75*2**(self.level+1) > 0:
			self._level += 1;

	@property
	def money(self):
		return(self._money);
	
	@money.setter
	def money(self, new_money):
		self._money = new_money;
	
	#===========================================================

	def credit(self, how_much, constant_scr):
		"""Check if the credit is correct
		Return True if it is, else return False """
		if(how_much > 0):
			if self.money + how_much >= 0:
				self.money += how_much;
				self.display_budget(constant_scr);
				return(True);
	
		return(False);

	#===========================================================

	def debit(self, how_much, constant_scr):
		if(how_much > 0):
			if (self.money - how_much) >= 0:
				self.money -= how_much;
				self.display_budget(constant_scr);
				return(True);

		return(False);
	
	#===========================================================

	def display_const(self, const_scr):
		self.display_budget(const_scr);
		const_scr.addstr(1, 18, "Niveau {0}".format(self.level));
		const_scr.addstr(2, 18, "{0} xp manquant pour".format(str(75*2**(self.level+1) - self.xp)));
		const_scr.addstr(3, 18, "passer au niveau supérieur");
		const_scr.refresh();

	#===========================================================

	def display_budget(self, const_scr):
		const_scr.addstr(1, 1, "Jour numéro {0}".format(self.days));
		const_scr.addstr(2, 1, "{0} : {1}".format(emoji.emojize(":euro_banknote:"), self.money));
		const_scr.refresh();

	#===========================================================
	
	def do_price_list(self, ingredients_buying_price_list):
		price_list = [];
		for i, j in ingredients_buying_price_list:
			if (i in self.ingredients.keys()):
				price_list.append("   Déjà acheté   ");

			else:
				len_price = len(str(j)) + 1;
				temp = " " * ((17 - len_price) // 2);
				temp += str(j) + "€";
				temp += " " * (17 - len(temp));
				price_list.append(temp);
		return (price_list);
	
	#===========================================================

	def upgrade_menu(self, main_scr, const_scr):
		main_scr.clear();
		main_scr.addstr(2, 2, "Que voulez-vous faire ?");
		main_scr.addstr(3, 4, "1 : Acheter de nouveaux ingrédients");
		main_scr.addstr(4, 4, "2 : Améliorer votre four");

		key = main_scr.getkey();

		while (key not in '12'):
			key = main_scr.getkey();

		if (key == '1'):
			self.upgrade_ingredient(main_scr, const_scr);
		
		else:
			self.upgrade_equipment(main_scr, const_scr);

	#===========================================================
	
	def upgrade_ingredient(self, main_scr, const_scr):
		main_scr.clear();
		with open("../packages/ingredients_buying_price.json", "r") as f:
			ingredients_buying_price = json.load(f);

		with open("../languages/french/connection_emoji_ingredient.json", "r") as f:
			connection_emoji_ingredient = json.load(f);

		emoji_list = ["        " + emoji.emojize(":"+i+":") + "       " for i in ingredients_buying_price.keys()];
		
		ingredient_name_list = [];
		k = 1;
		for i in ingredients_buying_price.keys():
			name = connection_emoji_ingredient[i];
			temp = " " * ((17 - len(name)) // 2 - 1);
			temp += str(k) + ": " + name;
			temp += " " * (17 - len(temp));
			ingredient_name_list.append(temp);
			k += 1;

		ingredients_buying_price_list = list(ingredients_buying_price.items());
		
		key = 'e';
		scr_number = 0; #Screen number for the ingredients
		price_list = self.do_price_list(ingredients_buying_price_list);
		while (key != 'q'):
			temp = "{"+str(0+scr_number)+"}|{"+str(1+scr_number)+"}|{"+str(2+scr_number)+"}|{"+str(3+scr_number)+"}";
			main_scr.addstr(2, 2, "   " + temp.format(*emoji_list) + "   ");
			main_scr.addstr(3, 2, "<--" + temp.format(*ingredient_name_list) + "-->");
			main_scr.addstr(4, 2, "   " + temp.format(*price_list) + "   ");
			main_scr.addstr(5, 2, "Appuyez sur le numéro de l'ingrédient que vous voulez acheter, \nou 'q' pour quitter");
			main_scr.refresh();
			key = main_scr.getkey().lower();

			if (key == 'key_right'):
				scr_number = scr_number + 4 if scr_number + 7 < len(emoji_list)-1 else len(emoji_list) - 4;

			elif (key == 'key_left'):
				scr_number = scr_number - 4 if scr_number - 4 > 0 else 0;

			elif (key == 'q'):
				break;

			else:
				with open("../packages/all_ingredients.json", "r") as f:
					temp = json.load(f);
		
				if (self.debit(ingredients_buying_price_list[int(key)-1][1], const_scr)):
					self.ingredients[ingredients_buying_price_list[int(key)-1][0]] = temp[ingredients_buying_price_list[int(key)-1][0]];
					price_list = self.do_price_list(ingredients_buying_price_list);

		main_scr.clear();
		main_scr.refresh();
	
	#===========================================================

	def upgrade_equipment(self, main_scr, const_scr):
		main_scr.clear();
		key = 'e';

		while (key != 'q'):
			main_scr.addstr(2, 2, "Équipement : Four");
			main_scr.addstr(3, 2, "Niveau : " + str(self.oven_level));
			main_scr.addstr(4, 2, "Temps de cuisson : " + str(self.cooking_time) + " secondes");
			main_scr.addstr(5, 2, "Gain de l'amélioration : 0.2 secondes");
			main_scr.addstr(6, 2, "Prix : 100€");
			main_scr.addstr(7, 2, "Appuyez sur 1 pour améliorer, ou 'q' pour quitter");
			main_scr.refresh();

			key = main_scr.getkey().lower();
			
			if (key == '1'):
				if (self.debit(100, const_scr)):
					self.oven_level += 1;
					self.cooking_time = round(self.cooking_time - 0.2, 2);

		main_scr.clear();
		main_scr.refresh();
			



core/Pizza.py
# Pizza file
"""
Pizza documentation

Properties: 
>>> Client client : "Client object that command the pizza"
>>> window scr : "Main screen (where the pizza is cooked)"
>>> window constant_scr : "Screen that show money left, number of the day, level, ..."
>>> Restaurant restaurant : "Restaurant object (the player)"
>>> list ingredient_wanted : "List of wanted ingredients"
>>> dict ingredients_on_pizza : "Dict of ingredients currently on the pizza ({ingredient_name: number_of_that})"
>>> list ingredients_coordinates : "List of coordinates of every ingredient ([[x, y], [x, y], ...])"
>>> int x : "x of the cursor"
>>> int y : "y of the cursor"
>>> bool cooked : "if the pizza is cooked or not"
>>> int outgoings_expense : "sum of outgoings expense"

Methods: 
>>> __init__(self, Client client, Restaurant restaurant, list ingredient_wanted, window scr, window constant_scr) : "same as properties"
	constructor of the pizza
>>> do_pizza(self)
	do what is needed to cook the pizza
>>> display_pizza(self, int color_pair) : "color_pair: number of which color pair for displaying"
	display the pizza
>>> correct_position(self)
	check if the player can add an ingredient at the current position
>>> overdraft(self, str ingredient) : "ingredient: name of the ingredient that is added"
	check if the player have enough money and debit the player if that's the case
>>> put_ingredients(self)
	method that recover key that are entered, move the cursor and serve the client at the end
>>> add_ingredient(self, str which_one) : "which_one: ingredient that whants to be added"
	add the ingredient if it can be
>>> cook(self)
	cook the pizza
"""
# Global import
import curses;
import time;
import math;
import emoji;

class Pizza:
	def __init__(self, client, restaurant, ingredient_wanted, scr, constant_scr):
		self.client = client;
		self.scr = scr;
		self.constant_scr = constant_scr;
		self.restaurant = restaurant;
		self.ingredient_wanted = ingredient_wanted;
		self.ingredients_on_pizza = {}; # dict : {ingredient_name: number}
		self.ingredients_coordinates = []; # list of coordinates of every : [[x, y], ...]
		self.y = 8;
		self.x = 18;
		self.cooked = False;
		self.outgoings_expense = 0;

	#======================================================================
	
	def do_pizza(self):
		self.scr.clear();
		#Debit base price
		if (not self.overdraft(2)):
			#Display of the pizza
			self.display_pizza_edge(4);
			curses.curs_set(2);
			#Put ingredients on the pizza
			self.put_ingredients();

	#======================================================================

	def display_pizza(self, color_pair):
		self.scr.addstr(3,  14, ", - ~~ - ,", curses.color_pair(color_pair));
		
		self.scr.addstr(4,  10, ", '", curses.color_pair(color_pair));
		self.scr.addstr(4,  25, "' ,", curses.color_pair(color_pair));

		self.scr.addstr(5,  8,  ",", curses.color_pair(color_pair));
		self.scr.addstr(5,  29, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(6,  7,  ",", curses.color_pair(color_pair));
		self.scr.addstr(6,  30, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(7,  6,  ",", curses.color_pair(color_pair));
		self.scr.addstr(7,  31, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(8,  6,  ",", curses.color_pair(color_pair));
		self.scr.addstr(8,  31, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(9,  6,  ",", curses.color_pair(color_pair));
		self.scr.addstr(9,  31, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(10, 7,  ",", curses.color_pair(color_pair));
		self.scr.addstr(10, 30, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(11, 8,  ",", curses.color_pair(color_pair));
		self.scr.addstr(11, 29, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(12, 10, ",", curses.color_pair(color_pair));
		self.scr.addstr(12, 28, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(13, 12, "' - , __ , - '", curses.color_pair(color_pair));
		
		self.scr.refresh();

	#======================================================================

	def display_pizza_edge(self, color_pair):
		self.scr.addstr(2, 13,  ", -  ~~  - ,", curses.color_pair(color_pair));
		
		self.scr.addstr(3, 9,  ", '", curses.color_pair(color_pair));
		self.scr.addstr(3, 26,  "' ,", curses.color_pair(color_pair));
		
		self.scr.addstr(4,  7,  ",", curses.color_pair(color_pair));
		self.scr.addstr(4, 30,  ",", curses.color_pair(color_pair));
		
		self.scr.addstr(5,  6,  ",", curses.color_pair(color_pair));
		self.scr.addstr(5, 31,  ",", curses.color_pair(color_pair));
		
		self.scr.addstr(6,  5,  ",", curses.color_pair(color_pair));
		self.scr.addstr(6, 32,  ",", curses.color_pair(color_pair));
		
		self.scr.addstr(7,  5,  ",", curses.color_pair(color_pair));
		self.scr.addstr(7, 32,  ",", curses.color_pair(color_pair));
		
		self.scr.addstr(8,  5,  ",", curses.color_pair(color_pair));
		self.scr.addstr(8, 32,  ",", curses.color_pair(color_pair));
		
		self.scr.addstr(9,  5,  ",", curses.color_pair(color_pair));
		self.scr.addstr(9, 32,  ",", curses.color_pair(color_pair));
		
		self.scr.addstr(10, 5,  ",", curses.color_pair(color_pair));
		self.scr.addstr(10, 32, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(11, 6,  ",", curses.color_pair(color_pair));
		self.scr.addstr(11, 31, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(12, 7,  ",", curses.color_pair(color_pair));
		self.scr.addstr(12, 30, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(13, 9,  ",", curses.color_pair(color_pair));
		self.scr.addstr(13, 28, ",", curses.color_pair(color_pair));
		
		self.scr.addstr(14, 11, "' - ,  __  , - '", curses.color_pair(color_pair));
		
		self.scr.move(self.y, self.x);
		self.scr.refresh();

	#======================================================================

	def correct_position(self):
		dx = abs(17 - self.x);
		dy = abs(8 - self.y)*2.5;
		d = math.sqrt(dx**2 + (dy**2));

		if(d < 11):		
			for i in self.ingredients_coordinates:
				if (i[0] == self.x and i[1] == self.y):
					return(False);

			return(True);

		else:
			return(False);
	
	#======================================================================

	def overdraft(self, price):
		if(self.restaurant.debit(price, self.constant_scr)):
			self.outgoings_expense += price;
			return(False);

		else:
			return(True);

	#======================================================================

	def put_ingredients(self):
		key = self.scr.getkey().lower();

		while(key.lower() != 'q'):
			#Movements
			if(key == 'key_right'):
				self.old_x = self.x;
				self.x += 2 if self.x <= 37 else 0;

			elif(key == 'key_left'):
				self.old_x = self.x;
				self.x -= 2 if self.x >= 1 else 0;

			elif(key == 'key_up'):
				self.old_y = self.y;
				self.y -= 1 if self.y >= 1 else 0;

			elif(key == 'key_down'): 
				self.old_y = self.y;
				self.y += 1 if self.y <= 18 else 0; 

			# Check if player want to cook the pizza
			elif(key == 'e'):
				self.cook();

			elif(key == 't'):
				if(not self.overdraft(self.restaurant.ingredients["tomato"] * 8)):
					self.display_pizza(1);
					try:
						self.ingredients_on_pizza["tomato"] += 8;
					except KeyError:
						self.ingredients_on_pizza["tomato"] = 8;

			#Put an ingredient
			elif(key in self.restaurant.connection_key_ingredient.keys()):
				self.add_ingredient(self.restaurant.connection_key_ingredient[key]);

			
			self.scr.move(self.y, self.x);
			key = self.scr.getkey().lower();
	
		self.client.served();

	#======================================================================

	def add_ingredient(self, which_one):
		if(self.correct_position() and not self.overdraft(self.restaurant.ingredients[which_one])):		
			self.scr.addstr(self.y, self.x, emoji.emojize(":"+which_one+":"));
			self.ingredients_coordinates.append([self.x, self.y]);

			try:
				self.ingredients_on_pizza[which_one] += 1;
			except KeyError:
				self.ingredients_on_pizza[which_one] = 1;

			self.scr.refresh();
	
	#======================================================================

	def cook(self):
		time.sleep(self.restaurant.cooking_time);
		self.cooked = True;
		self.display_pizza_edge(3);
		try:
			if (self.ingredients_on_pizza["tomato"] >= 1):
				self.display_pizza(1);
		except KeyError:
			pass;
